engine/src/vault/view.rs

```diff
@@ -9,11 +9,39 @@ use crate::vault::{
     },
 };
 
+use runtime::GuardedVec;
 use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, convert::Infallible, fmt::Debug};
+use thiserror::Error as DeriveError;
 
-use std::collections::HashMap;
+use super::crypto_box::DecryptError;
 
-use runtime::GuardedVec;
+#[derive(DeriveError, Debug)]
+pub enum VaultError<TProvErr: Debug, TProcErr: Debug = Infallible> {
+    #[error("Vault `{0:?}` does not exist.")]
+    VaultNotFound(VaultId),
+
+    #[error("Record Error: `{0:?}`")]
+    Record(#[from] RecordError<TProvErr>),
+
+    #[error("Procedure Error `{0:?}`")]
+    Procedure(TProcErr),
+}
+
+#[derive(DeriveError, Debug)]
+pub enum RecordError<TProvErr: Debug> {
+    #[error("Provider Error: `{0:?}`")]
+    Provider(TProvErr),
+
+    #[error("Found Invalid Transaction")]
+    CorruptedContent,
```

Comment:
Since this is basically a fatal error, it would be helpful for it to contain more information. Alternatively, since fatal errors require human intervention, it also seems okay to log additional information.